juju/juju
# [JUJU-875] Remove use of "Force=true" when removing an application

Remove "Force=true" in garbageCollectOneApplication when removing an application and clean up properly.

We believe we had originally added this "destroyOp.Force = true" to get something going quickly initially, but it's not appropriate to force when the user hasn't requested it.

Once we removed this we also had to do the following so the units and application get removed correctly:

* Clear the "has-resources" flag once the K8s resources have been deleted. This is done in the caasapplicationprovisioner application worker's dead() method.
* Also in caasapplicationprovisioner, watch the application's units and call RemoveUnit when a unit is dead (for pod-spec charms, this is done in worker/caasoperator, but we weren't doing it for sidecar charms).

I've added unit tests for all the new code paths, plus  a basic integration test for `juju remove-application --force`, and we already have an integration test for `remove-application` without force (https://github.com/juju/juju/pull/13929).

The commit https://github.com/juju/juju/pull/13973/commits/5ebc199b2de89872ad053f2a82d46b0447a5faf0 should also fix the race condition that causes an application not to be removed from state sometimes: https://bugs.launchpad.net/juju/+bug/1970508

## QA steps

See integration tests -- just `deploy` and `remove-application` on a sidecar application:

```
juju bootstrap microk8s
# optional logging for debugging:
# juju model-config logging-config="<root>=INFO;juju.worker.caasapplicationprovisioner=DEBUG;juju.state=DEBUG"
juju deploy snappass-test
# wait for it to deploy
juju remove-application snappass-test
juju status
```

To repro the shutdown race condition, do the following:

```
# bootstrap and add logging as per above
juju deploy snappass-test -n10
# wait for it to deploy
juju remove-application snappass-test
juju status
```

Every other time or so with 10 units it will not delete the application from state due to the race condition mentioned in https://github.com/juju/juju/pull/13973/commits/5ebc199b2de89872ad053f2a82d46b0447a5faf0.
